include/dali/core/mm/async_pool.h

```diff
@@ -0,0 +1,457 @@
+          sync_stream_ = CUDAStream::Create(true);
+        CUDA_DTOR_CALL(cudaStreamWaitEvent(sync_stream_, kv.second.free_list.head->event, 0));
+      }
+    }
+    if (sync_stream_)
+      CUDA_DTOR_CALL(cudaStreamSynchronize(sync_stream_));
+  }
+
+  void *do_allocate(size_t bytes, size_t alignment) override {
+    if (device_id_ == -1)
+      CUDA_CALL(cudaGetDevice(&device_id_));
+    adjust_size_and_alignment(bytes, alignment);
+    std::lock_guard<LockType> guard(lock_);
+    return allocate_from_global_pool(bytes, alignment);
+  }
+
+  void do_deallocate(void *mem, size_t bytes, size_t alignment) override {
+    adjust_size_and_alignment(bytes, alignment);
+    std::lock_guard<LockType> guard(lock_);
+    char *ptr = static_cast<char *>(mem);
+    pop_block_padding(ptr, bytes, alignment);
+    return global_pool_.deallocate(ptr, bytes, alignment);
+  }
+
+  /**
+   * @brief Tries to recycle per-stream free memory or allocates from the global pool.
+   *
+   * There are two main cases:
+   * (1) Attempt to allocate from stream-specific free list - a smallest suitable block is used,
+   *     padding is applied, if necessary, and recorded.
+   * (2) Otherwise, the requested memory is allocated from global pool.
+   *
+   * Per-stream frees are not coalesced - therefore, even if the total memory freed on a stream
+   * may be sufficient, there may be no satisfactory block.
+   */
+  void *do_allocate_async(size_t bytes, size_t alignment, stream_view stream) override {
+    if (device_id_ == -1)
+      CUDA_CALL(cudaGetDevice(&device_id_));
+    adjust_size_and_alignment(bytes, alignment);
+    std::lock_guard<LockType> guard(lock_);
+    auto it = stream_free_.find(stream.value());
+    void *ptr;
+    if (it != stream_free_.end()) {
+      ptr = try_allocate(it->second, bytes, alignment);
+      if (ptr)
+        return ptr;
+    }
+    return allocate_from_global_pool(bytes, alignment);
+  }
+
+  /**
+   * @brief Allocates from a global pool, possibly releasing per-stream memory to the global pool.
+   *
+   * (1) If the global pool can't allocate enough memory without allocating upstream blocks,
+   *     and there are some per-stream pending frees, the completed per-streem frees are returned
+   *     to the global pool.
+   * (2) Then, if there are still some pending frees and `avoid_upstream` option was selected at
+   *     construction
+   * (3)    Try to allocate, again avoiding upstream - on failure
+   * (4)    Synchronize and reclaim pending frees
+   * (5) Allocate from the global pool.
+   */
+  void *allocate_from_global_pool(size_t bytes, size_t alignment) {
+    void *ptr;
+    if (num_pending_frees_ == 0) {
+      ptr = global_pool_.allocate(bytes, alignment);
+      return ptr;
+    }
+    ptr = global_pool_.try_allocate_from_free(bytes, alignment);
+    if (ptr)
+      return ptr;
+    for (auto &kv : stream_free_)
+      free_pending(kv.second);
+    if (avoid_upstream_ && num_pending_frees_ > 0) {
```

Comment:
```suggestion
    if (avoid_upstream_) {
```
`free_pending` just decreases `num_pending_frees_`.